### DISCUSSION TITLE
Confused by API when using across two applications on same domain

### DISCUSSION THREAD

**Discussion by mildfuzz** [vvo/iron-session#848] [Q&A]
I have two applications (`myApp1` and `myApp2`)

They share secrets with a shared vault store, both apps run on the same domain (`myApp1` at https://myapp.com/, `myApp2` at  https://myapp.com/myApp1)

User first goes to `myApp1`, which handles logging them in and populating some session data, they are then sent to `myApp2`

The trouble I have is when I try to read the session from `myApp2` using `getIronSession`, the value is undefined. BUT when I use `unsealData` to access the value from the cookie, I can see values.

I feel like I am misunderstanding the intention of the library here. Any assistance would be great :)

**Comment by mildfuzz** [comment]
Edited to add more context

**Comment by mildfuzz** [comment]
So it appears when I run it locally, it works as expected

**Comment by vvo** [comment]
@mildfuzz it would be tremendous to have a github repository I can run to help you out, for example a create-next-app simple repo https://nextjs.org/docs/pages/api-reference/cli/create-next-app with the structure of /app1 /app2 and what you expect to happen in more details. thanks!

**Reply by mildfuzz** [reply]
okay, so further investigation and it might not be so trivial.  If I run locally by following following procedure:
- Run `myApp1` on `localhost:3000`, populate session
- Kill `myApp1`
- Run `myApp2` on localhost:3000/myApp2
I can see that `getIronSession` works as expected

however, when I run in an environment, I can only get the values using `unsealData`

I have tried running both apps in prod mode locally to see if that's the different, but it works in that scenario too.

So there is *something* different in the deployed version that affects this and I am struggling to identify it.

**Reply by mildfuzz** [reply]
So raising dummy repos might not solve this...

**Comment by vvo** [comment]
@mildfuzz v9 can tell you what is going on here. When a cookie cannot be unsealed, iron-session starts a fresh empty session instead of throwing, which is exactly the "getIronSession is empty but unsealData sees values" shape you hit. Add `onUnsealError`:

```ts
const options = {
  cookieName: "session",
  password: process.env.SESSION_PASSWORD,
  onUnsealError: (reason, error) => console.warn(reason, error),
};
```

`"unknown-password"` means myApp2's password does not match myApp1's, `"expired"` means the two apps disagree on `ttl`, and `"invalid"` means the value is not a seal at all. If nothing fires and the session is still empty, the cookie is not reaching myApp2, so check that both apps use the same `cookieName` and that the cookie's `path` is `/`.

Given it works locally and not deployed, my guess is the password differs between the two deployed apps. Tell me the reason you get if it is none of those.

v9 notes: https://github.com/vvo/iron-session/releases/tag/v9.0.0
